### ISSUE TITLE
Thanks!

### ISSUE THREAD

**Issue by coderschoolreview** [singuyenmai/Machine-Learning-CoderSchool#1]
The goal of this assignment was to introduce you to 2 main concepts in Machine Learning: Data Pre-processing, and Classification. You learned how to query and clean data using the pandas library in Python, and built a simple Machine Learning Classifier based on the K Nearest Neighbors algorithm.

Amazing job with the assignment. You've done a great job documenting your code and explaining what you're doing, and best of all: explaining why you think you saw a certain result, such as how `dropna` can actually hurt your accuracy.

Keep up the good work with the homework assignments. Good job exploring the interesting machine learning questions.

**Comment by coderschoolreview** [comment]
**Assignment 2**

The goal of this assignment was to introduce you to three new classification techniques and to understand how to select the best parameters and features for them. You learned how to use python built-in functions (GridSearchCV, SelectKBest, RFE, SelectFromModel) to try out new models (Support Vector Machines, Random Forests, and Logistic Regression) and test different permutations of parameter values and features, and analyze your results to help build better machine learning models.

Good job!

Here's what you did well:
* Very modular code -- you write functions and then use those functions to re-use and re-run many different combinations, such as putting all your params/estimators in a dictionary and then passing those objects to functions. Good design!
* Nice job investigating the effect of not normalizing the data. As you noticed, it most strongly affects classifiers that work on distance metrics (egs: `SVC`).
* You seem comfortable with Python and programming in general. Your code is solid. 

Some areas to improve on:
* You are missing the confusion matrix and classification report for all optimized classifiers.
* You did not discuss your results. It is important to talk about your results and evaluate them so you can demonstrate that you are understanding the concepts being studied. Example:
    * How did the optimized classifiers work compared to the original and best?
    * Did `GridSearchCV` always give you good results? Why? Or why not?
* Looks like you got the following error:
`ConvergenceWarning: The max_iter was reached which means the coef_ did not converge`
You can try fixing this by increasing the `max_iter` parameter in `GridSearchCV` (the default is 100).
* Next time, I would encourage you to try at least one of the bonus stories, such as adding `moods` (like you did for Assignment 1).

Overall, good work. Keep it up!

**Comment by coderschoolreview** [comment]
**Assignment 3**

The goal of this assignment was to introduce you to three new Natural Language Processing techniques, and to understand how to perform some basic sentiment analysis on song lyrics using these methods. You learned how to clean and prepare textual information for NLP, and then apply the following approaches: Bag Of Words, TF-IDF, and Doc2Vec. You used your prior knowledge of Python estimators, feature selection, and parameter optimization techniques to produce feature vectors from these NLP methods to make predictions on the moods of songs using their lyrics.

Great job!! I am impressed with how you always modularize your code by writing functions to do all the work. It's the sign of clean design and thinking like a programmer :)

What you did well:
* Good job exploring new classifiers, like the Multinomial Naive Bayes! As you might have discovered, Multinomial NB can be a good choice when working with NLP. Here is a great article on it: 
https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/
* Nice plots of accuracy! Really helps to be able to visually see how classifiers are performing.

Some feedback:
* The error you got with `Features [0 0 0 ..., 0 0 0] are constant` : not sure, this might be because there's a mistake somewhere in your code, or maybe there is actually a feature that is constant and does not change (i.e. has no variance). In this case you can try using `VarianceThreshold` first. Documentation is here:
http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.VarianceThreshold.html

* I'm curious to see if the 'noise' caused by n-grams happens with just the NB classifier or with other classifiers too?  Perhaps try investigating to see how other classifiers / moods work with n-grams.

Overall, excellent work :)
